src/dstack/_internal/server/services/offers.py

```diff
@@ -92,4 +100,48 @@ async def get_offers_by_requirements(
     if profile.instance_types is not None:
         offers = [(b, o) for b, o in offers if o.instance.name in profile.instance_types]
 
-    return offers
+    if blocks is None or blocks == 1:
+        return offers
+
+    shareable_offers = []
+    for backend, offer in offers:
+        resources = offer.instance.resources
+        gpus = resources.gpus
+        if not gpus:
+            continue
+        if gpus[0].vendor == gpuhunt.AcceleratorVendor.GOOGLE:
+            # TPU instances cannot be shared
+            continue
+        if blocks == "auto":
+            divisor = len(gpus)
+        else:
+            divisor = blocks
+        if len(gpus) % divisor or resources.cpus % divisor:
+            # gpus or cpus are not divisible by blocks
+            continue
+        shareable_offers.append((backend, offer))
```

Comment:
I think this code can be nicely refactored into a separate func.

Reply:
Refactored in https://github.com/dstackai/dstack/pull/2253/commits/8b555f7a6ad65afbed244e3d1e76359bdbbcbbf9